contracts/multihop/src/tests/swap.rs

```diff
@@ -3,158 +3,363 @@ use crate::storage::Swap;
+    let swap3 = Swap {
+        offer_asset: token3.address.clone(),
+        ask_asset: token4.address.clone(),
     };
 
-    factory_client.create_liquidity_pool(&first_lp_init_info);
-    factory_client.create_liquidity_pool(&second_lp_init_info);
-    factory_client.create_liquidity_pool(&third_lp_init_info);
-
-    // 3. provide liquidity for each one of the liquidity pools
-    for lp in factory_client.query_pools() {
-        let lp_client = lp_contract::Client::new(&env, &lp);
-        lp_client.provide_liquidity(
-            &user.clone(),
-            &Some(1_000_000i128),
-            &None,
-            &Some(1_000_000i128),
-            &None,
-            &None::<i64>,
-        );
-    }
+    let operations = vec![&env, swap1, swap2, swap3];
+
+    multihop.swap(&recipient, &operations, &5_000i128);
+
+    // 5. check if it goes according to plan
+    assert_eq!(token1.balance(&recipient), 0i128);
+    assert_eq!(
+        token4.balance(&recipient),
+        4_956i128,
+        "token4 not as expected"
+    );
+}
+
+#[test]
+fn swap_three_different_pools_with_fees() {
+    let env = Env::default();
+
+    let admin = Address::random(&env);
+
+    env.mock_all_auths();
+    env.budget().reset_unlimited();
+
+    let token1 = deploy_and_mint_tokens(&env, &admin, 10_000_000i128);
+    let token2 = deploy_and_mint_tokens(&env, &admin, 10_000_000i128);
+    let token3 = deploy_and_mint_tokens(&env, &admin, 10_000_000i128);
+    let token4 = deploy_and_mint_tokens(&env, &admin, 10_000_000i128);
+
+    // 1. deploy factory
+    let factory_client = deploy_and_initialize_factory(&env, admin.clone());
+
+    deploy_and_initialize_lp(
+        &env,
+        &factory_client,
+        admin.clone(),
+        token1.address.clone(),
+        1_000_000,
+        token2.address.clone(),
+        1_000_000,
+        Some(1_000),
+    );
+    deploy_and_initialize_lp(
+        &env,
+        &factory_client,
+        admin.clone(),
+        token2.address.clone(),
+        2_000_000,
+        token3.address.clone(),
+        2_000_000,
+        Some(1_000),
+    );
+    deploy_and_initialize_lp(
+        &env,
+        &factory_client,
+        admin.clone(),
+        token3.address.clone(),
+        3_000_000,
+        token4.address.clone(),
+        3_000_000,
+        Some(1_000),
+    );
 
     // 4. swap with multihop
-    let multihop = deploy_multihop_contract(&env, admin, &factory_client.address);
+    let multihop = deploy_multihop_contract(&env, admin.clone(), &factory_client.address);
     let recipient = Address::random(&env);
-    tokens[0].mint(&recipient, &50i128);
-    assert_eq!(tokens[0].balance(&recipient), 50i128);
-    assert_eq!(tokens[3].balance(&recipient), 0i128);
+    token1.mint(&recipient, &10_000i128);
+    assert_eq!(token1.balance(&recipient), 10_000i128);
+    assert_eq!(token2.balance(&recipient), 0i128);
+    assert_eq!(token3.balance(&recipient), 0i128);
+    assert_eq!(token4.balance(&recipient), 0i128);
```

Comment:
Let's test what needs to be tested, we don't need to check that tokens that weren't minted before aren't there.

```suggestion
```